storage/block device mounts (#16488)
docker/docs: storage/volumes.md

	--name cif-volume
```

The `addr` option is required if you specify a hostname instead of an IP.
This lets Docker perform the hostname lookup.

### Mount a block storage device

You can mount a block storage device, such as an external drive or a drive partition, to a container.
The following example shows how to create and use a file as a block storage device,

> and how to mount the file as a container volume.

Author: Should we refer to the `mount` syscall, or should we instead say that we're shelling out to the `mount` program, and refer to the [mount](https://linux.die.net/man/8/mount) documentation for the relevant kernel version for information about supported filesystem options?

> and how to mount the file as a container volume.

Author: Added a note that precedes this step list. PTAL

> and how to mount the file as a container volume.

Reviewer: Technically we don't shell out and use the syscall directly. But I think we could use the mount command example to illustrate the equivalent behavior, as the reader is more likely to know how to use the mount command and not the syscall 😄 

See my comment on the note below.

> and how to mount the file as a container volume.

Reviewer: ```suggestion
and how to mount the block device as a container volume.
```

1. Create a file and allocate some space to it:

   ```console
   $ fallocate -f 1G disk.raw
   ```

2. Build a filesystem onto the `disk.raw` file:

   ```console
   $ mkfs.ext4 disk.raw
   ```

3. Create a loop device for the file:

   ```console
   $ losetup -f --show disk.raw
   /dev/loop5
   ```

4. Create a volume that uses the file as a device block mount:

   ```console
   $ docker volume create my_block_device \
     -o 'device=/dev/loop5' \
     -o 'type=ext4'
   my_block_device

>    ```

Author: Updated to fully use `--mount` options as opposed to creating a named volume

	--name cif-volume
```

The `addr` option is required if you specify a hostname instead of an IP.
This lets Docker perform the hostname lookup.

### Mount a block storage device

You can mount a block storage device, such as an external drive or a drive partition, to a container.
The following example shows how to create and use a file as a block storage device,
and how to mount the file as a container volume.

> **Note**
>
> The following procedure is only an example. This example creates an `ext4`

> > filesystem and mounted into a container. The filesystem support of your

Reviewer: ```suggestion
> filesystem and mounts it into a container. The filesystem support of your
```

> system depends on the version of the Linux kernel you are using.
>

> > Under the hood, the `--mount` flag invokes the Linux `mount` command and forwards

Reviewer: ```suggestion
Under the hood, the `--mount` flag invokes the Linux `mount` syscall and forwards
```

Reviewer: Also, to be technically correct - this only happens for the `local` driver, not the `--mount` flag itself.

> the options you pass to it unaltered. Docker doesn't implement any additional

> > functionality on top of the native mount feature supported by the Linux kernel.

Reviewer: I think we could use the `mount` command call to illustrate what happens:

```bash
mount -t <mount.volume-opt.type> <mount.volume-opt.device> <mount.dst> -o <mount.volume-opts.o>
```

so the example
```bash
--mount='type=volume,dst=/external-drive,volume-driver=local,volume-opt=device=/dev/loop5,volume-opt=type=ext4'
```

is basically an equivalent to:
```bash
mount -t ext4 /dev/loop5 /external-drive
```

1. Create a file and allocate some space to it:

   ```console
   $ fallocate -f 1G disk.raw
   ```

2. Build a filesystem onto the `disk.raw` file:

   ```console
   $ mkfs.ext4 disk.raw
   ```

3. Create a loop device:

   ```console
   $ losetup -f --show disk.raw
   /dev/loop5
   ```

   > **Note**
   >
   > `losetup` creates an ephemeral loop device that's removed after

>    > system reboot, or manually removed with `losetup -d`.

Reviewer: Maybe we should also add a final step which does the cleanup with `losetup -d /dev/loop5`?

That's why we're using the `--mount` option for the `docker run` command instead.

#### Example: Mounting a block device in a container

The following steps create an `ext4` filesystem and mounts it into a container.
The filesystem support of your system depends on the version of the Linux kernel you are using.

1. Create a file and allocate some space to it:

   ```console
   $ fallocate -f 1G disk.raw
   ```

2. Build a filesystem onto the `disk.raw` file:

   ```console
   $ mkfs.ext4 disk.raw
   ```

3. Create a loop device:

   ```console
   $ losetup -f --show disk.raw
   /dev/loop5
   ```

   > **Note**
   >
   > `losetup` creates an ephemeral loop device that's removed after
   > system reboot, or manually removed with `losetup -d`.

4. Run a container that mounts the loop device as a volume:

   ```console
   $ docker run -it --rm \
     --mount='type=volume,dst=/external-drive,volume-driver=local,volume-opt=device=/dev/loop5,volume-opt=type=ext4' \
     ubuntu bash
   ```

   When the container starts, the path `/external-drive` mounts the
   `disk.raw` file from the host filesystem as a block device.

5. When you're done, and the device is unmounted from the container,

>    detach the loop device to remove the mount from the host system:

Reviewer: nit: `remove the mount` -> `remove the loop device` (there's no mount on the host system, only the device :))

This lets Docker perform the hostname lookup.

### Mount a block storage device

You can mount a block storage device, such as an external drive or a drive partition, to a container.
The following example shows how to create and use a file as a block storage device,
and how to mount the block device as a container volume.

> **Important**
>
> The following procedure is only an example.
> The solution illustrated here isn't recommended as a general practice.
> Don't attempt this approach unless you're very confident about what you're doing.
{: .important }

Under the hood, the `--mount` flag using the `local` storage driver invokes the
Linux `mount` syscall and forwards the options you pass to it unaltered.
Docker doesn't implement any additional functionality on top of the native mount features supported by the Linux kernel.

If you're familiar with the
[Linux `mount` command](https://man7.org/linux/man-pages/man8/mount.8.html),
you can think of the `--mount` options being forwarded to the `mount` command as follows:

```console
$ mount -t <mount.volume-opt.type> <mount.volume-opt.device> <mount.dst> -o <mount.volume-opts.o>
```

> In other words, the following commands are equivalent:

Reviewer: "Equivalent" is a bit too strong IMO. It sounds like these two commands would produce exactly the same result, but they don't.
First runs a container that mounts it, and second mount it on the host.

It would be more correct to say, that the from the container perspective, docker run example sets up the mount in a way that executing "mount -t ext4 /dev/loop5 /external-drive" inside container would (if the container had direct access to /dev/loop5, but it doesn't, so you must rely on `docker run` to do that).

```console
$ docker run \
  --mount='type=volume,dst=/external-drive,volume-driver=local,volume-opt=device=/dev/loop5,volume-opt=type=ext4'

$ mount -t ext4 /dev/loop5 /external-drive
```

The following steps create an `ext4` filesystem and mounts it into a container.
The filesystem support of your system depends on the version of the Linux kernel you are using.

1. Create a file and allocate some space to it:

   ```console
   $ fallocate -f 1G disk.raw
   ```

2. Build a filesystem onto the `disk.raw` file:

   ```console
   $ mkfs.ext4 disk.raw
   ```

3. Create a loop device:

   ```console
   $ losetup -f --show disk.raw
   /dev/loop5
   ```

   > **Note**
   >
   > `losetup` creates an ephemeral loop device that's removed after
   > system reboot, or manually removed with `losetup -d`.

4. Run a container that mounts the loop device as a volume:

   ```console
   $ docker run -it --rm \
     --mount='type=volume,dst=/external-drive,volume-driver=local,volume-opt=device=/dev/loop5,volume-opt=type=ext4' \
     ubuntu bash
   ```

   When the container starts, the path `/external-drive` mounts the
   `disk.raw` file from the host filesystem as a block device.

> 5. When you're done, detach the loop device to remove the mount
>    from the host system.

Reviewer: I'd phrase it more like this, because we didn't mount this device on host:

```suggestion
5. When you're done and the device is unmounted from the container,
   detach the loop device from the host system.
```
